### PR TITLE
[PLANNER-1821] Convert CoachShuttleGathering example to constraint streams

### PR DESCRIPTION
**Pull request by cuijulian**
Jira: https://issues.redhat.com/browse/PLANNER-1821

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    private Constraint coachStopLimit(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BusStop.class)
+                .filter(busStop -> busStop.getBus() instanceof Coach)
+                .groupBy(busStop -> (Coach) busStop.getBus(), count())
+                .filter((coach, stopCount) -> stopCount > coach.getStopLimit())
+                .penalizeLong("coachStopLimit", HardSoftLongScore.ofUninitialized(0, 1000000, 0),
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java]: Why not `HardSoftLongScore.ofHard(1000000)`?
Similarly in the other constraints.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                .filter(busStop -> busStop.getTransportTimeToHub() != null && busStop.getTransportTimeRemainder() < 0)
+                .penalizeLong("transportTime", HardSoftLongScore.ONE_HARD, busStop -> -1 * busStop.getTransportTimeRemainder());
+    }
+
+    private Constraint shuttleDestinationIsCoachOrHub(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shuttle.class)
+                .filter(shuttle -> shuttle.getDestination() != null)
+                .join(StopOrHub.class, equal(Shuttle::getDestination, stopOrHub -> stopOrHub))
+                .filter((shuttle, stopOrHub) -> !stopOrHub.isVisitedByCoach())
+                .penalize("shuttleDestinationIsCoachOrHub", HardSoftLongScore.ofUninitialized(0, 1000000000, 0));
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java]: For these very large numbers, I suggest you use `_` to separate thousands. `1000000000` becomes `1_000_000_000`.

**Comment by triceo** [comment]
Marking as stale, we do not expect to merge this before the code style changes are merged.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        return constraintFactory.from(BusStop.class)
+                .filter(busStop -> busStop.getBus() instanceof Shuttle)
+                .groupBy(BusStop::getBus, sum(BusStop::getPassengerQuantity))
+                .filter((shuttle, requiredCapacity) -> requiredCapacity > shuttle.getCapacity())
+                .penalizeLong("shuttleCapacity", HardSoftLongScore.ofHard(1000),
+                        (shuttle, requiredCapacity) -> requiredCapacity - shuttle.getCapacity());
+    }
+
+    // TODO: Implement this CS
+    private Constraint coachCapacity(ConstraintFactory constraintFactory) {
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java:75]: This is the painful constraint

**Pull request closed without merging**